### ISSUE TITLE
Dtls cannot establish with libdatachannel

### ISSUE THREAD

**Issue by giangndm-bluesea** [algesten/str0m#261]
When we work with OBS broadcast (which uses libdatachannel internally), DTLS connection cannot be established.

```
[2023-08-07T10:12:18.325Z INFO  str0m::ice::agent] Add local candidate: Candidate(host=192.168.66.223:64225 prio=2130706175)
[2023-08-07T10:12:18.325Z TRACE str0m::ice::agent] Calculated local preference: 65534
[2023-08-07T10:12:18.327Z INFO  str0m::ice::agent] Set remote credentials: IceCreds { ufrag: "p5IV", pass: "MeL2sgkoqVHqORLqUmVP9W" }
[2023-08-07T10:12:18.327Z INFO  str0m::ice::agent] Add remote candidate: Candidate(host=192.168.66.223:53735 prio=2122317567)
[2023-08-07T10:12:18.327Z TRACE str0m::ice::agent] Form pair local: Candidate(host=192.168.66.223:64225 prio=2130706175) remote: Candidate(host=192.168.66.223:53735 prio=2122317567)
[2023-08-07T10:12:18.327Z DEBUG str0m::ice::agent] Add new pair CandidatePair(0-0 prio=9115284546252701182 state=Waiting attempts=0 unanswered=0 remote=0 last=None nom=None)
[2023-08-07T10:12:18.327Z INFO  str0m::ice::agent] Add remote candidate: Candidate(host=10.243.178.47:53735 prio=2122317823)
[2023-08-07T10:12:18.327Z TRACE str0m::ice::agent] Form pair local: Candidate(host=192.168.66.223:64225 prio=2130706175) remote: Candidate(host=10.243.178.47:53735 prio=2122317823)
[2023-08-07T10:12:18.327Z DEBUG str0m::ice::agent] Add new pair CandidatePair(0-1 prio=9115285645764328958 state=Waiting attempts=0 unanswered=0 remote=0 last=None nom=None)
[2023-08-07T10:12:18.327Z INFO  str0m] DTLS setup is: false
[2023-08-07T10:12:18.327Z DEBUG str0m::session] Enable TWCC feedback
[2023-08-07T10:12:18.327Z INFO  str0m::media] Set Mid(0) extension map: Extensions()
[2023-08-07T10:12:18.327Z INFO  str0m::change::sdp] Adding SSRC: SsrcInfo { ssrc: Ssrc(5002), repairs: None, cname: Some("mjRNPIvR0kqaQS4q"), stream_id: Some("ZETlb4lFx3Lb2skD"), track_id: Some("ZETlb4lFx3Lb2skD-audio") }
[2023-08-07T10:12:18.327Z DEBUG str0m::streams::receive] Create StreamRx for SSRC: 5002
[2023-08-07T10:12:18.327Z INFO  str0m::media] Set Mid(1) extension map: Extensions()
[2023-08-07T10:12:18.327Z INFO  str0m::change::sdp] Adding SSRC: SsrcInfo { ssrc: Ssrc(5000), repairs: None, cname: Some("mjRNPIvR0kqaQS4q"), stream_id: Some("ZETlb4lFx3Lb2skD"), track_id: Some("ZETlb4lFx3Lb2skD-video") }
[2023-08-07T10:12:18.327Z DEBUG str0m::streams::receive] Create StreamRx for SSRC: 5000
[2023-08-07T10:12:18.330Z TRACE str0m::ice::agent] Enqueueing event: IceRestart(IceCreds { ufrag: "jz1H", pass: "7F517yJ2CxkpnlaC8VMsLf" })
[2023-08-07T10:12:18.330Z INFO  str0m::ice::agent] State change (new connection): New -> Checking
[2023-08-07T10:12:18.330Z TRACE str0m::ice::agent] Enqueueing event: IceConnectionStateChange(Checking)
[2023-08-07T10:12:18.330Z TRACE str0m::ice::agent] Handle next triggered pair: CandidatePair(0-1 prio=9115285645764328958 state=Waiting attempts=0 unanswered=0 remote=0 last=None nom=None)
[2023-08-07T10:12:18.330Z TRACE str0m::ice::pair] Check state: Waiting -> InProgress
[2023-08-07T10:12:18.330Z DEBUG str0m::ice::agent] Send STUN request: 192.168.66.223:64225 -> 10.243.178.47:53735 StunMessage { method: Binding, class: Request, attrs: [Username("p5IV:jz1H"), IceControlled(8986517625568425202), Priority(1862270719), MessageIntegrityMark, FingerprintMark], integrity_len: 0 }
[2023-08-07T10:12:18.343Z DEBUG str0m::streams::receive] Created feedback RR: ReceiverReport { sender_ssrc: Ssrc(3354681331), reports: [] }
[2023-08-07T10:12:18.343Z DEBUG str0m::streams::receive] Created feedback extended receiver report: ExtendedReport { ssrc: Ssrc(5002), blocks: [Rrtr(Rrtr { ntp_time: MediaTime(3900391938331714, 1000000) })] }
[2023-08-07T10:12:18.343Z DEBUG str0m::streams::receive] Created feedback RR: ReceiverReport { sender_ssrc: Ssrc(3354681331), reports: [] }
[2023-08-07T10:12:18.343Z DEBUG str0m::streams::receive] Created feedback extended receiver report: ExtendedReport { ssrc: Ssrc(5000), blocks: [Rrtr(Rrtr { ntp_time: MediaTime(3900391938331714, 1000000) })] }
[2023-08-07T10:12:18.343Z TRACE str0m::ice::agent] Poll event: Some(IceRestart(IceCreds { ufrag: "jz1H", pass: "7F517yJ2CxkpnlaC8VMsLf" }))
[2023-08-07T10:12:18.343Z TRACE str0m::ice::agent] Poll event: Some(IceConnectionStateChange(Checking))
[2023-08-07T10:12:18.343Z DEBUG str0m] IceConnectionStateChange(Checking)
[2023-08-07T10:12:18.343Z TRACE str0m::ice::agent] Poll transmit: Transmit { source: 192.168.66.223:64225, destination: 10.243.178.47:53735, len: 88 }
[2023-08-07T10:12:18.343Z TRACE str0m] OUT Transmit { source: 192.168.66.223:64225, destination: 10.243.178.47:53735, len: 88 }
[2023-08-07T10:12:18.343Z TRACE str0m] IN Receive { source: 192.168.66.223:53735, destination: 192.168.66.223:64225, contents: Stun(StunMessage { method: Binding, class: Request, attrs: [Priority(1853882367), IceControlling(6386116986444904902), Unknown(32802), Username("jz1H:p5IV"), MessageIntegrity(_), Fingerprint(1634438764)], integrity_len: 68 }) }
[2023-08-07T10:12:18.343Z TRACE str0m::ice::agent] Handle receive: Receive { source: 192.168.66.223:53735, destination: 192.168.66.223:64225, contents: Stun(StunMessage { method: Binding, class: Request, attrs: [Priority(1853882367), IceControlling(6386116986444904902), Unknown(32802), Username("jz1H:p5IV"), MessageIntegrity(_), Fingerprint(1634438764)], integrity_len: 68 }) }
[2023-08-07T10:12:18.343Z TRACE str0m::ice::agent] Check if accepts message: StunMessage { method: Binding, class: Request, attrs: [Priority(1853882367), IceControlling(6386116986444904902), Unknown(32802), Username("jz1H:p5IV"), MessageIntegrity(_), Fingerprint(1634438764)], integrity_len: 68 }
[2023-08-07T10:12:18.343Z TRACE str0m::ice::agent] Message accepted
[2023-08-07T10:12:18.343Z TRACE str0m::ice::agent] Remote candidate for STUN request found
[2023-08-07T10:12:18.343Z TRACE str0m::ice::agent] Found existing pair for STUN request: CandidatePair(0-0 prio=9115284546252701182 state=Waiting attempts=0 unanswered=0 remote=0 last=None nom=None)
[2023-08-07T10:12:18.343Z TRACE str0m::ice::pair] Remote binding requests: 1
[2023-08-07T10:12:18.343Z DEBUG str0m::ice::agent] Send STUN reply: 192.168.66.223:64225 -> 192.168.66.223:53735 StunMessage { method: Binding, class: Success, attrs: [XorMappedAddress(192.168.66.223:53735), MessageIntegrityMark, FingerprintMark], integrity_len: 0 }
[2023-08-07T10:12:18.343Z TRACE str0m::ice::agent] Enqueueing event: DiscoveredRecv { source: 192.168.66.223:53735 }
[2023-08-07T10:12:18.343Z TRACE str0m::ice::agent] Poll event: Some(DiscoveredRecv { source: 192.168.66.223:53735 })
[2023-08-07T10:12:18.343Z INFO  str0m] ICE remote address: 192.168.66.223:53735
[2023-08-07T10:12:18.344Z TRACE str0m::ice::agent] Poll transmit: Transmit { source: 192.168.66.223:64225, destination: 192.168.66.223:53735, len: 64 }
[2023-08-07T10:12:18.344Z TRACE str0m] OUT Transmit { source: 192.168.66.223:64225, destination: 192.168.66.223:53735, len: 64 }
[2023-08-07T10:12:18.344Z TRACE str0m] IN Receive { source: 192.168.66.223:53735, destination: 192.168.66.223:64225, contents: Stun(StunMessage { method: Binding, class: Request, attrs: [Priority(1853882111), UseCandidate, IceControlling(6386116986444904902), Unknown(32802), Username("jz1H:p5IV"), MessageIntegrity(_), Fingerprint(3002203556)], integrity_len: 72 }) }
[2023-08-07T10:12:18.344Z TRACE str0m::ice::agent] Handle receive: Receive { source: 192.168.66.223:53735, destination: 192.168.66.223:64225, contents: Stun(StunMessage { method: Binding, class: Request, attrs: [Priority(1853882111), UseCandidate, IceControlling(6386116986444904902), Unknown(32802), Username("jz1H:p5IV"), MessageIntegrity(_), Fingerprint(3002203556)], integrity_len: 72 }) }
[2023-08-07T10:12:18.344Z TRACE str0m::ice::agent] Check if accepts message: StunMessage { method: Binding, class: Request, attrs: [Priority(1853882111), UseCandidate, IceControlling(6386116986444904902), Unknown(32802), Username("jz1H:p5IV"), MessageIntegrity(_), Fingerprint(3002203556)], integrity_len: 72 }
[2023-08-07T10:12:18.344Z TRACE str0m::ice::agent] Message accepted
[2023-08-07T10:12:18.344Z TRACE str0m::ice::agent] Binding request sent USE-CANDIDATE
[2023-08-07T10:12:18.344Z TRACE str0m::ice::agent] Remote candidate for STUN request found
[2023-08-07T10:12:18.344Z TRACE str0m::ice::agent] Found existing pair for STUN request: CandidatePair(0-0 prio=9115284546252701182 state=Waiting attempts=0 unanswered=0 remote=1 last=Some(Instant { tv_sec: 1147222, tv_nsec: 571870958 }) nom=None)
[2023-08-07T10:12:18.348Z TRACE str0m::ice::pair] Remote binding requests: 2
[2023-08-07T10:12:18.348Z DEBUG str0m::ice::pair] Nominated pair: CandidatePair(0-0 prio=9115284546252701182 state=Waiting attempts=0 unanswered=0 remote=2 last=Some(Instant { tv_sec: 1147222, tv_nsec: 572741375 }) nom=Nominated)
[2023-08-07T10:12:18.348Z DEBUG str0m::ice::agent] Send STUN reply: 192.168.66.223:64225 -> 192.168.66.223:53735 StunMessage { method: Binding, class: Success, attrs: [XorMappedAddress(192.168.66.223:53735), MessageIntegrityMark, FingerprintMark], integrity_len: 0 }
[2023-08-07T10:12:18.348Z TRACE str0m::ice::agent] Poll transmit: Transmit { source: 192.168.66.223:64225, destination: 192.168.66.223:53735, len: 64 }
[2023-08-07T10:12:18.348Z TRACE str0m] OUT Transmit { source: 192.168.66.223:64225, destination: 192.168.66.223:53735, len: 64 }
[2023-08-07T10:12:18.348Z TRACE str0m] IN Receive { source: 192.168.66.223:53735, destination: 192.168.66.223:64225, contents: Dtls(len: 342) }
[2023-08-07T10:12:18.348Z DEBUG str0m::dtls::ossl] Accept
[2023-08-07T10:12:18.381Z INFO  str0m::ice::agent] State change (got nomination, still trying others): Checking -> Connected
[2023-08-07T10:12:18.381Z TRACE str0m::ice::agent] Enqueueing event: IceConnectionStateChange(Connected)
[2023-08-07T10:12:18.381Z TRACE str0m::ice::agent] Nominating best candidate
[2023-08-07T10:12:18.381Z TRACE str0m::ice::agent] Enqueueing event: NominatedSend { source: 192.168.66.223:64225, destination: 192.168.66.223:53735 }
[2023-08-07T10:12:18.381Z TRACE str0m::ice::agent] Handle next triggered pair: CandidatePair(0-0 prio=9115284546252701182 state=Waiting attempts=0 unanswered=0 remote=2 last=Some(Instant { tv_sec: 1147222, tv_nsec: 572741375 }) nom=Nominated)
[2023-08-07T10:12:18.381Z DEBUG str0m::ice::pair] Nominated attempt STUN binding: CandidatePair(0-0 prio=9115284546252701182 state=Waiting attempts=0 unanswered=0 remote=2 last=Some(Instant { tv_sec: 1147222, tv_nsec: 572741375 }) nom=Nominated)
[2023-08-07T10:12:18.381Z TRACE str0m::ice::pair] Check state: Waiting -> InProgress
[2023-08-07T10:12:18.381Z DEBUG str0m::ice::agent] Send STUN request: 192.168.66.223:64225 -> 192.168.66.223:53735 StunMessage { method: Binding, class: Request, attrs: [Username("p5IV:jz1H"), IceControlled(8986517625568425202), Priority(1862270719), MessageIntegrityMark, FingerprintMark], integrity_len: 0 }
[2023-08-07T10:12:18.381Z TRACE str0m::ice::agent] Poll event: Some(IceConnectionStateChange(Connected))
[2023-08-07T10:12:18.381Z DEBUG str0m] IceConnectionStateChange(Connected)
[2023-08-07T10:12:18.381Z TRACE str0m::ice::agent] Poll event: Some(NominatedSend { source: 192.168.66.223:64225, destination: 192.168.66.223:53735 })
[2023-08-07T10:12:18.381Z INFO  str0m] ICE nominated send from: 192.168.66.223:64225 to: 192.168.66.223:53735
[2023-08-07T10:12:18.381Z TRACE str0m::ice::agent] Poll transmit: Transmit { source: 192.168.66.223:64225, destination: 192.168.66.223:53735, len: 88 }
[2023-08-07T10:12:18.381Z TRACE str0m] OUT Transmit { source: 192.168.66.223:64225, destination: 192.168.66.223:53735, len: 88 }
[2023-08-07T10:12:18.381Z TRACE str0m::dtls] Poll datagram: 256
[2023-08-07T10:12:18.381Z TRACE str0m] OUT Transmit { source: 192.168.66.223:64225, destination: 192.168.66.223:53735, len: 256 }
[2023-08-07T10:12:18.381Z TRACE str0m::dtls] Poll datagram: 256
[2023-08-07T10:12:18.381Z TRACE str0m] OUT Transmit { source: 192.168.66.223:64225, destination: 192.168.66.223:53735, len: 256 }
[2023-08-07T10:12:18.383Z TRACE str0m::dtls] Poll datagram: 256
[2023-08-07T10:12:18.383Z TRACE str0m] OUT Transmit { source: 192.168.66.223:64225, destination: 192.168.66.223:53735, len: 256 }
[2023-08-07T10:12:18.383Z TRACE str0m::dtls] Poll datagram: 256
[2023-08-07T10:12:18.383Z TRACE str0m] OUT Transmit { source: 192.168.66.223:64225, destination: 192.168.66.223:53735, len: 256 }
[2023-08-07T10:12:18.383Z TRACE str0m::dtls] Poll datagram: 256
[2023-08-07T10:12:18.383Z TRACE str0m] OUT Transmit { source: 192.168.66.223:64225, destination: 192.168.66.223:53735, len: 256 }
[2023-08-07T10:12:18.383Z TRACE str0m::dtls] Poll datagram: 61
[2023-08-07T10:12:18.383Z TRACE str0m] OUT Transmit { source: 192.168.66.223:64225, destination: 192.168.66.223:53735, len: 61 }
[2023-08-07T10:12:18.383Z TRACE str0m] IN Receive { source: 192.168.66.223:53735, destination: 192.168.66.223:64225, contents: Stun(StunMessage { method: Binding, class: Success, attrs: [XorMappedAddress(192.168.66.223:64225), Unknown(32802), MessageIntegrity(_), Fingerprint(2584379561)], integrity_len: 44 }) }
[2023-08-07T10:12:18.383Z TRACE str0m::ice::agent] Handle receive: Receive { source: 192.168.66.223:53735, destination: 192.168.66.223:64225, contents: Stun(StunMessage { method: Binding, class: Success, attrs: [XorMappedAddress(192.168.66.223:64225), Unknown(32802), MessageIntegrity(_), Fingerprint(2584379561)], integrity_len: 44 }) }
[2023-08-07T10:12:18.383Z TRACE str0m::ice::agent] Check if accepts message: StunMessage { method: Binding, class: Success, attrs: [XorMappedAddress(192.168.66.223:64225), Unknown(32802), MessageIntegrity(_), Fingerprint(2584379561)], integrity_len: 44 }
[2023-08-07T10:12:18.383Z TRACE str0m::ice::agent] Message accepted
[2023-08-07T10:12:18.383Z TRACE str0m::ice::agent] Found local candidate for mapped address: 192.168.66.223:64225
[2023-08-07T10:12:18.383Z DEBUG str0m::ice::pair] Nomination success: CandidatePair(0-0 prio=9115284546252701182 state=InProgress attempts=1 unanswered=0 remote=2 last=Some(Instant { tv_sec: 1147222, tv_nsec: 572741375 }) nom=Success)
[2023-08-07T10:12:18.383Z TRACE str0m::ice::pair] Check state: InProgress -> Succeeded
[2023-08-07T10:12:18.383Z TRACE str0m::ice::pair] Recorded binding response: CandidatePair(0-0 prio=9115284546252701182 state=Succeeded attempts=1 unanswered=0 remote=2 last=Some(Instant { tv_sec: 1147222, tv_nsec: 572741375 }) nom=Success)
[2023-08-07T10:12:18.383Z TRACE str0m] IN Receive { source: 192.168.66.223:53735, destination: 192.168.66.223:64225, contents: Dtls(len: 15) }
[2023-08-07T10:12:18.383Z DEBUG str0m::dtls::ossl] DTLS failure: Error { code: ErrorCode(1), cause: Some(Ssl(ErrorStack([Error { code: 336602130, library: "SSL routines", function: "dtls1_read_bytes", reason: "sslv3 alert bad certificate", file: "ssl/record/rec_layer_d1.c", line: 615, data: "SSL alert number 42" }]))) }
```

May be have some issues with ossl.rs.
Ps: I also tested with webrtc-rs and it work well.

**Comment by algesten** [comment]
@giangndm-bluesea thanks for that.

It might be the expected SSL algorithm isn't there. Can you post the SDP?

**Comment by giangndm-bluesea** [comment]
@algesten Bellow is Offer from Libdatachannel and Answer from Str0m

```
v=0
    o=rtc 3479147068 0 IN IP4 127.0.0.1
    s=-
    t=0 0
    a=group:BUNDLE 0 1
    a=group:LS 0 1
    a=msid-semantic:WMS *
    a=setup:actpass
    a=ice-ufrag:ooFK
    a=ice-pwd:Giv0kf3WBRsvl/NpomCICy
    a=ice-options:ice2,trickle
    a=fingerprint:sha-256 EF:95:E2:3C:39:5F:6C:0B:D9:B9:1C:87:29:22:77:D1:91:34:0D:CB:F0:92:69:C5:A2:72:92:06:98:F8:AA:85
    m=audio 56530 UDP/TLS/RTP/SAVPF 111
    c=IN IP4 10.243.178.47
    a=mid:0
    a=sendonly
    a=ssrc:446539624 cname:HIw16uf5wUgY2R7K
    a=ssrc:446539624 msid:DqqsVyiXSR1mlGpQ DqqsVyiXSR1mlGpQ-audio
    a=msid:DqqsVyiXSR1mlGpQ DqqsVyiXSR1mlGpQ-audio
    a=rtcp-mux
    a=rtpmap:111 opus/48000/2
    a=fmtp:111 minptime=10;maxaveragebitrate=96000;stereo=1;sprop-stereo=1;useinbandfec=1
    a=candidate:1 1 UDP 2122317823 10.243.178.47 56530 typ host
    a=candidate:2 1 UDP 2122317567 192.168.66.223 56530 typ host
    a=end-of-candidates
    m=video 56530 UDP/TLS/RTP/SAVPF 96
    c=IN IP4 10.243.178.47
    a=mid:1
    a=extmap:1 urn:ietf:params:rtp-hdrext:sdes:mid
    a=extmap:2 urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id
    a=sendonly
    a=msid:DqqsVyiXSR1mlGpQ DqqsVyiXSR1mlGpQ-video
    a=rid:l send
    a=rid:m send
    a=rid:h send
    a=simulcast:send l;m;h
    a=rtcp-mux
    a=rtpmap:96 H264/90000
    a=rtcp-fb:96 nack
    a=rtcp-fb:96 nack pli
    a=rtcp-fb:96 goog-remb
    a=fmtp:96 profile-level-id=42e01f;packetization-mode=1;level-asymmetry-allowed=1
```

```
    v=0
    o=str0m-0.1.0 3801435588447092337 2 IN IP4 0.0.0.0
    s=-
    t=0 0
    a=group:BUNDLE 0 1
    a=msid-semantic: WMS DqqsVyiXSR1mlGpQ
    m=audio 9 UDP/TLS/RTP/SAVPF 111
    c=IN IP4 0.0.0.0
    a=candidate:8402408543720050545 1 udp 2130706175 192.168.66.223 55734 typ host ufrag x2sb
    a=ice-ufrag:x2sb
    a=ice-pwd:1vDMMeruMKTl62Ygvzsm1O
    a=ice-options:trickle
    a=fingerprint:sha-256 FF:74:71:9C:BD:A3:3A:42:34:F2:DB:A2:EA:CD:2B:2A:28:1E:D4:C9:C4:81:E0:4C:19:F8:C7:BB:43:97:B0:6C
    a=setup:passive
    a=mid:0
    a=recvonly
    a=msid:DqqsVyiXSR1mlGpQ DqqsVyiXSR1mlGpQ-audio
    a=rtcp-mux
    a=rtpmap:111 opus/48000/2
    a=rtcp-fb:111 transport-cc
    a=fmtp:111 minptime=10;useinbandfec=1
    m=video 9 UDP/TLS/RTP/SAVPF 96
    c=IN IP4 0.0.0.0
    a=ice-ufrag:x2sb
    a=ice-pwd:1vDMMeruMKTl62Ygvzsm1O
    a=ice-options:trickle
    a=fingerprint:sha-256 FF:74:71:9C:BD:A3:3A:42:34:F2:DB:A2:EA:CD:2B:2A:28:1E:D4:C9:C4:81:E0:4C:19:F8:C7:BB:43:97:B0:6C
    a=setup:passive
    a=mid:1
    a=extmap:1 urn:ietf:params:rtp-hdrext:sdes:mid
    a=extmap:2 urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id
    a=recvonly
    a=msid:DqqsVyiXSR1mlGpQ DqqsVyiXSR1mlGpQ-video
    a=rtcp-mux
    a=rtpmap:96 H264/90000
    a=rtcp-fb:96 transport-cc
    a=rtcp-fb:96 goog-remb
    a=rtcp-fb:96 ccm fir
    a=rtcp-fb:96 nack
    a=rtcp-fb:96 nack pli
    a=fmtp:96 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f
    a=rid:l recv
    a=rid:m recv
    a=rid:h recv
    a=simulcast:recv l;m;h
```

**Comment by giangndm-bluesea** [comment]
For more information, I attached Wireshark captured of the failed case

[libdatachannel-connect-error.pcapng.zip](https://github.com/algesten/str0m/files/12282665/libdatachannel-connect-error.pcapng.zip)

**Comment by algesten** [comment]
Hi @giangndm-bluesea it seems libdatachannel is rejecting our self generated TLS cert for some unknown reason (42). Could you provide the log that libdatachannel produces during this?

Also do you have some test code you can share?

**Comment by giangndm-bluesea** [comment]
@algesten It only happened with OBS WHIP build version, (we can test by using not released version at https://github.com/obsproject/obs-studio/actions/runs/5815986617 ).

When I try with libdatachannel direct it does not occur above issue. But when switch to using libnice for ICE, which failed to connect but I think it is another issue.

If have any update about that error, I will update soon.

**Comment by jellevd246** [comment]
I have the same problem. 
And after some testing it seems that adding some issuer and subject information to the certificate solves the problem.

```rust
let mut x509_name = openssl::x509::X509NameBuilder::new().unwrap();
x509_name.append_entry_by_text("C", "US").unwrap();
x509_name.append_entry_by_text("ST", "CA").unwrap();
x509_name.append_entry_by_text("O", "Some organization").unwrap();
x509_name.append_entry_by_text("CN", "www.example.com").unwrap();
let x509_name = x509_name.build();

x509b.set_subject_name(&x509_name)?;
x509b.set_issuer_name(&x509_name)?;
```
Put this here, but maybe with relevant information.
https://github.com/algesten/str0m/blob/28956b093ac6d8b29127f455d7352fa88e52b67a/src/dtls/ossl.rs#L54

**Comment by algesten** [comment]
Nice catch! Let's fix that!

**Comment by algesten** [comment]
This is fixed by #402

**Comment by algesten** [comment]
@jellevd246 or @giangndm-bluesea can anyone confirm this is fixed in the main branch now?

**Comment by jellevd246** [comment]
@algesten Yes OBS WHIP has no problem connecting now for me.

**Comment by algesten** [comment]
🎉 

Thank you both so much for investigating this! I'll make a patch release now.

**Comment by algesten** [comment]
Released in 0.4.1
